src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h

```diff
@@ -0,0 +1,161 @@
+		 * cost_function */
+		FirstOrderSAGCostFunctionInterface(
+		    SGMatrix<float64_t>* X, SGMatrix<float64_t>* y,
+		    Matrix<var, Dynamic, 1>* trainable_parameters,
+		    Matrix<function<var(int32_t)>, Dynamic, 1>* cost_for_ith_point,
+		    function<var(Matrix<var, Dynamic, 1>*)>* total_cost);
+
+		/** Default constructor, use setter helpers to set X, y,
+		*trainable_parameters, and
+		*		cost_function
+		*/
+		FirstOrderSAGCostFunctionInterface(){};
+
+		/** Setter for the training data X */
+		virtual void set_training_data(
+		    SGMatrix<float64_t>* X_new, SGMatrix<float64_t>* y_new);
+
+		/** Setter for the trainable parameters of the cost function */
+		virtual void
+		    set_trainable_parameters(Matrix<var, Dynamic, 1>* new_params);
+
+		/** Setter for the cost function definition using stan */
+		virtual void set_ith_cost_function(
+		    Matrix<function<var(int32_t)>, Dynamic, 1>* new_cost_f);
+
+		/** Setter for the overall cost function */
+		virtual void set_cost_function(
+		    function<var(Matrix<var, Dynamic, 1>*)>* total_cost);
+
+		virtual ~FirstOrderSAGCostFunctionInterface();
+
+		/** Initialize to generate a sample sequence
+		 *
+		 */
+		virtual void begin_sample();
+
+		/** Get next sample
+		 *
+		 * @return false if reach the end of the sample sequence
+		 * */
+		virtual bool next_sample();
+
+		/** Get the SAMPLE gradient value wrt target variables
+		 *
+		 * WARNING
+		 * This method does return
+		 * \f$ \frac{\partial f_i(w) }{\partial w} \f$,
+		 * instead of
+		 * \f$\sum_i{ \frac{\partial f_i(w) }{\partial w} }\f$
+		 *
+		 * For least squares cost function, that is the value of
+		 * \f$\frac{\partial f_i(w) }{\partial w}\f$ given \f$w\f$ is known
+		 * where the index \f$i\f$ is obtained by next_sample()
+		 *
+		 * @return sample gradient of variables
+		 */
+		virtual SGVector<float64_t> get_gradient();
+
+		/** Get the cost given current target variables
+		 *
+		 * For least squares, that is the value of \f$f(w)\f$.
+		 *
+		 * @return cost
+		 */
+		virtual float64_t get_cost();
+
+		/** Get the sample size
+		 *
+		 * @return the sample size
+		 */
+		virtual index_t get_sample_size();
+
+		/** Get the AVERAGE gradient value wrt target variables
+		 *
+		 * Note that the average gradient is the mean of sample gradient from
+		 * get_gradient()
+		 * if samples are generated (uniformly) at random.
+		 *
+		 * WARNING
+		 * This method returns
+		 * \f$ \frac{\sum_i^n{ \frac{\partial f_i(w) }{\partial w} }}{n}\f$
+		 *
+		 * For least squares, that is the value of
+		 * \f$ \frac{\frac{\partial f(w) }{\partial w}}{n} \f$ given \f$w\f$ is
+		 * known
+		 * where \f$f(w)=\frac{ \sum_i^n{ (y_i-w^t x_i)^2 } }{2}\f$
+		 *
+		 * @return average gradient of target variables
+		 */
+		virtual SGVector<float64_t> get_average_gradient();
+
+	protected:
+		/** X is the training data in column major matrix format */
+		SGMatrix<float64_t>* m_X;
```

Comment:
Why are all these members pointers?

Reply:
@iglesias Just a convention (I try to pass things by reference/pointer instead of value to avoid copying). But I agree I will get rid of SGMatrix's pointers since it already implements that under the hood.